### PR TITLE
Add smoothing

### PR DESCRIPTION
**Pull request by drew-sinha**
First attempt at adding smoothing with a trailing moving average; feel free to hack away/give comments as appropriate. I'll also touch base about the TODO that I mention in the comment/commit.

### CODE REVIEW

**Comment by drew-sinha** [comment]
Does this look better for the factory function? I refactored to bring all of the masking for appropriate timepoints to smooth over into the filter function.

**Comment by drew-sinha** [comment]
Here's round 3. I think I got mostly everything down.... Let me know if I missed anything else.

File: elegant/worm_data.py [hunk 1]
```diff
@@ -441,6 +441,78 @@ def interpolate_feature(self, feature, age, age_feature='age'):
         ages, value = self.get_time_range(feature, age_feature=age_feature)
         return numpy.interp(age, ages, value)
 
+    def make_smoothing_filter(filter_type, **filter_params):
```
Comment (zpincus) [inline elegant/worm_data.py]: Probably better for this to be a helper function (outside of the class), with a _ prepended to the name. As is, it's not a staticmethod nor a self parameter, so it currently won't work as intended...

File: elegant/worm_data.py [hunk 2]
```diff
@@ -441,6 +441,78 @@ def interpolate_feature(self, feature, age, age_feature='age'):
...
+            return numpy.dot(1/numpy.exp(-t**2/(2*sigma)).sum() * numpy.exp(-t**2/(2*sigma)), data[smoothing_mask])
+        def uniform_filter(time, data, window_length):
+            smoothing_mask = (time>=0) & (time <= window_length)
+            return numpy.dot(smoothing_mask/smoothing_mask.sum(),data)
+
+        filters = {'gaussian': gaussian_filter,
+            'uniform': uniform_filter}
+
+        def smoothing_filter(time, data):
+            selected_filter = filters[filter_type](**filter_params)
```
Comment (zpincus) [inline elegant/worm_data.py]: This won't work either.
```
def smoothing_filter(time, data):
    return filters[filter_type](time, data, **filter_params)
```
is what you want.

File: elegant/worm_data.py [hunk 3]
```diff
@@ -441,6 +441,78 @@ def interpolate_feature(self, feature, age, age_feature='age'):
...
+        """Smooth a feature with a prescribed filter
+
+            Parameters
+                feature - feature to filter over
+                filter_type - str corresponding to the filter to be applied ('gaussian', 'uniform'),
+                    or a callable with the call signature smoothing_filter(time,data)
+                age_feature: the name of the feature to use to determine the age
+                    window. Defaults to 'age', but other monotonic features, such as
+                    a time-left-alive 'ghost_age' could also be used. If this is a
+                    function, it will be called as age_feature(worm) to generate
```
Comment (zpincus) [inline elegant/worm_data.py]: There's a problem with `age_feature` being a function: `get_time_range` does the right thing, but later on you assume it's an attribute. Either get all the ages as `ages = age_feature(self) if callable(age_feature) else getattr(self.td, age_feature)` or don't advertise the possibility of it being callable.

My suspicion is that `get_time_range` shouldn't allow a callable either -- that just gets too confusing. I'll probably strip that out later.

File: elegant/worm_data.py [hunk 4]
```diff
@@ -441,6 +441,78 @@ def interpolate_feature(self, feature, age, age_feature='age'):
...
+                    to create filter.
+
+            The smoothed feature is the same size as the original feature and is
+                stored in the td attribute "smoothed_(feature)".
+        """
+
+        if callable(filter_type):
+            smoothing_filter = filter_type
+        else:
+            smoothing_filter = make_smoothing_filter(filter_type, window_length, **filter_params)
```
Comment (zpincus) [inline elegant/worm_data.py]: `window_length` isn't defined. That should just be one of the filter params now, no?

**Review by zpincus — APPROVED** [review]
Looks OK modulo my comments. I'm guessing you didn't test that the code actually works? It's OK, I've accidentally checked in some manifestly broken stuff too that you've caught :)

**Comment by drew-sinha** [comment]
Sounds good. I incorporated those last set of comments and tested the code; should be good for deployment.

**Comment by zpincus** [comment]
I made a few more changes after looking at the code: 2764dc3
What do you think? I think this is actually the simplest API of all...

I also clarified the docs and brought it in line with the new style guide.

**Comment by drew-sinha** [comment]
Left a couple of comments in comment

**Comment by zpincus** [comment]
Cool, thanks. My point with going through all these revisions to the API was just to model the cycle I usually take, BTW. I'll generally go through a few drafts of code before coming up with what seems maximally sensible. I think this converged on that, so thanks for bearing with...

**Pull request merged**